core/src/avm2/globals/flash/display3D/textures/texture.rs

```diff
@@ -1,8 +1,200 @@
+
+    if is_cube != atf_texture.cubemap {
+        return Err("Stage3D Texture and ATF Texture must both be cube/non-cube".into());
+    }
+
+    if atf_texture.width != texture.handle().width()
+        || atf_texture.height != texture.handle().height()
+    {
+        return Err("ATF texture dimensions do not match Texture dimensions".into());
+    }
+
+    // Just use the first mip level for now. We ignore the builtin format - the JPEG-XR format
+    // appears to override it
+    let mut first_mip = Cursor::new(&atf_texture.face_mip_data[0][0]);
+    let mut decoder =
+        jpegxr::ImageDecode::with_reader(&mut first_mip).expect("Failed to decode JPEG-XR image");
+
+    let pixel_format = decoder
+        .get_pixel_format()
+        .expect("Failed to get pixel format");
+    let (jpeg_width, jpeg_height) = decoder.get_size().expect("Failed to get JPEG-XR size");
+    let jpeg_width = jpeg_width as u32;
+    let jpeg_height = jpeg_height as u32;
+
+    assert_eq!(jpeg_width, atf_texture.width, "Mismatched JPEG-XR width");
+    assert_eq!(jpeg_height, atf_texture.height, "Mismatched JPEG-XR height");
+
+    let info = jpegxr::PixelInfo::from_format(pixel_format);
+    let stride = jpeg_width as usize * info.bits_per_pixel() / 8;
+    let size = stride * jpeg_height as usize;
+
+    // We convert the result to a TIFF - this makes the jpegxr library handle
+    // all of the weird JPEG-XR alpha formats for us. We can then use the normal
+    // `image` crate to decode the TIFF to an rgba array.
+    let mut bmp_buffer = vec![0; size];
+    decoder
+        .convert_to_tiff(&mut Cursor::new(&mut bmp_buffer))
+        .expect("Failed to convert to bitmap");
+
+    let image_reader =
+        image::io::Reader::with_format(Cursor::new(bmp_buffer), image::ImageFormat::Tiff);
+    let bitmap = image_reader
+        .decode()
+        .expect("Failed to decode Bitmap")
+        .to_rgba8();
+
+    // FIXME - are we handling premultiplied alpha correct?
+    let colors: Vec<_> = bitmap
+        .chunks_exact(4)
+        .map(|color| Color::argb(color[3], color[0], color[1], color[2]))
+        .collect();
+
+    let bitmap_data = BitmapData::new_with_pixels(
+        texture.handle().width(),
+        texture.handle().height(),
+        true,
+        colors,
+    );
+
+    let bitmap_data =
+        BitmapDataWrapper::new(GcCell::new(activation.context.gc_context, bitmap_data));
+
+    texture
+        .context3d()
+        .copy_bitmap_to_texture(bitmap_data.sync(), texture.handle(), 0);
+
+    Ok(())
+}
+
+pub fn upload_compressed_texture_from_byte_array<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    this: Object<'gc>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let texture = this.as_texture().unwrap();
+    let data = args.get_object(activation, 0, "data")?;
+    let byte_array_offset = args.get_u32(activation, 1)? as usize;
+    let async_ = args.get_bool(2);
```

Comment:
Should this be `_async` instead of `async_`?

Reply:
It's not really unused, but it conflicts with the `async` keyword